### PR TITLE
Keep track of transactions and threads associated with locks.

### PR DESCRIPTION
**Pull request by martinfurmanski**

### CODE REVIEW

File: community/kernel/src/main/java/org/neo4j/kernel/impl/locking/Locks.java [hunk 1]
```diff
@@ -107,6 +109,9 @@ public Factory( String key, String... altKeys )
         /** Release all locks. */
         void releaseAll();
 
+        /** Associate with transaction. */
+        void setTx( Transaction tx );
```
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/locking/Locks.java]: I'm cringing when looking at this setter. Why don't you make this association when creating/binding the client instead?

Comment (martinfurmanski) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/locking/Locks.java]: Cringe no more, I'll change it! Originally I was binding it directly to a thread, but transactions it seems can suspend/resume on another thread, so changed it to that... but with transactions a setter doesn't make the best sense anymore. I don't think there are any issues with changing it.

Comment (martinfurmanski) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/locking/Locks.java]: Actually, there seems to be a problem with that. The clients are pooled in forseti and even if we do not add the setTx to the Locks.Client interface, we would still need it on the ForsetiClient level, and the generalised flyweight pool would need a mechanism for abstractly initialising whatever it carries. Should be doable, but the change will be more complex.

**Comment by martinfurmanski** [comment]
@tinwelint You still want the change?

**Comment by chrisvest** [comment]
It'd be nice if the commit message said _why_ we'd do this change. I'd also rather the `Transaction` was added as a parameter to the `Locks.newClient()` method, rather than a setter on `Client`. This makes it more awkward to have this circular dependency, which leads me back to questioning why we do this.

**Comment by martinfurmanski** [comment]
It might be a bit awkward, but it's the minimal thing giving desired debugability on the maintenance branch. The awkwardness arises due to it being an afterthought and any major refactoring is not applicable for this branch at least, and not within the concept of a forward merge either.

**Comment by tinwelint** [comment]
I've done stuff like this locally when debugging, but to have it in the product is a differnt thing. Do we need it as part of production code?

**Comment by martinfurmanski** [comment]
We discussed it a bit offline. My view is that debugability is as important as any other feature of a product.

**Pull request merged**